#!/bin/bash
#$ -cwd -pe smp 6 -l mem=2G,time=1:: -N BamBWABam
# -cwd -l mem=1G,time=:10: -N BamBWABam
#This script takes a bam file and reverts it to sam format and then realigns with BWA mem
#	InpFil - (required) - Path to Bam file to be aligned. Alternatively a file with a list of bams can be provided and the script run in an array job. List file name must end ".list"
#	RefFiles - (required) - shell file to export variables with locations of reference files and resource directories; see list below
#	LogFil - (optional) - File for logging progress
#	PipeLine - P -(flag) - will start the GATK realign and recalibrate pipeline using the files generated by this script

#list of required reference files:
# $REF - reference genome in fasta format - must have been indexed using 'bwa index ref.fa'
# $EXOMPPLN - directory containing exome analysis pipeline scripts, 
#list of required tools:
# samtools <http://samtools.sourceforge.net/> <http://sourceforge.net/projects/samtools/files/>
# bwa mem <http://bio-bwa.sourceforge.net/> <http://sourceforge.net/projects/bio-bwa/files/>
# picard <http://picard.sourceforge.net/> <http://sourceforge.net/projects/picard/files/>
# HTSlib <https://github.com/samtools/htslib> <https://github.com/samtools/htslib/archive/master.zip>

## This file also require exome.lib.sh - which contains various functions used throughout my Exome analysis scripts; this file should be in the same directory as this script

###############################################################

#get arguments
PipeLine="false"
while getopts i:r:n:l:P opt; do
	case "$opt" in
		i) InpFil="$OPTARG";;
		r) RefFil="$OPTARG";; 
		n) BamNam="$OPTARG";; 
		l) LogFil="$OPTARG";;
		P) PipeLine="true";;
	esac
done

#load RefFil file
source $RefFil #load the references required
#Load script library
source $EXOMPPLN/exome.lib.sh

#set local variables
funcFilfromList
BamFil=`readlink -f $InpFil` #resolve absolute path to bam
BamNam=`basename $BamFil` 
BamNam=${BamNam/.bam/} # a name for the output files
if [[ -z $LogFil ]];then
	LogFil=$BamNam.BbB.log # a name for the log file
fi
AlnDir=$BamNam.align # directory in which processing will be done
AlnFil=$BamNam.bwamem.bam #filename for output file
DdpFil=$BamNam.bwamem.mkdup.bam #filename for file with PCR duplicates marked
mkdir -p $AlnDir # create working directory
cd $AlnDir # move into working directory
TmpLog=$LogFil.BbB.temp.log #use TmpLog for LogFil - for consistency with other exome analysis scripts

#start log
ProcessName="Align with BWA"
funcWriteStartLog
echo " Build of reference files: "$BUILD >> $TmpLog
echo "----------------------------------------------------------------" >> $TmpLog

#get ReadGroupHeader from input BAM
RgHeader=$(samtools view -H $BamFil | grep ^@RG | awk '{ gsub("\t","\\t") } { print }')
echo "ReadGroup header: $RgHeader" >> $TmpLog
if [[ $RgHeader == "" ]]||[[ $(echo "$RgHeader" | wc -l) -gt 1 ]]; then #check that we have a  RG header and if not write a warning to the log file
	echo "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++" >> $TmpLog
	echo "     Problem with ReadGroup header" >> $TmpLog
	echo "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++" >> $TmpLog
fi

###Align using BWA mem algorithm
StepName="Align with BWA mem"
StepCmd="htscmd bamshuf -uOn 128 $BamFil tmp | htscmd bam2fq -a - | gzip | bwa mem -M -R \"$RgHeader\" -t 6 -p $REF - | htscmd samview -bS - > $AlnFil"
#Stepcmd="bamshuf -uOn 128 aln_reads.bam tmp | htscmd bam2fq -a - | gzip > interleaved_reads.fq.gz"
#StepCmd="samtools bamshuf -Ou $BamFil | samtools bam2fq - | bwa mem -M -R \"$RgHeader\" -t 6 -p $REF - | samtools view -bS - > $AlnFil"
funcRunStep

#Final CleanUp
funcWriteEndLog
rm $TmpLog
